src/python/pants/backend/python/rules/python_test_runner.py

```diff
@@ -70,41 +49,24 @@ def run_python_test(test_target, pytest, python_setup, source_root_config, pex_b
     if hasattr(maybe_python_req_lib, 'requirements'):
       for py_req in maybe_python_req_lib.requirements:
         all_target_requirements.append(str(py_req.requirement))
-
-  # Sort all user requirement strings to increase the chance of cache hits across invocations.
-  all_requirements = sorted(all_target_requirements + list(pytest.get_requirement_strings()))
+  all_requirements = all_target_requirements + list(pytest.get_requirement_strings())
 
   # TODO(#7061): This str() can be removed after we drop py2!
   python_binary = text_type(sys.executable)
-  interpreter_constraint_args = parse_interpreter_constraints(
-    python_setup, python_target_adaptors=all_targets
-  )
-  interpreter_search_paths = text_type(create_path_env_var(python_setup.interpreter_search_paths))
+  interpreter_constraints = {
+    constraint
+    for target_adaptor in all_targets
+    for constraint in python_setup.compatibility_or_constraints(
+      getattr(target_adaptor, 'compatibility', None)
+    )
+  }
 
-  # TODO: This is non-hermetic because the requirements will be resolved on the fly by
-  # pex27, where it should be hermetically provided in some way.
   output_pytest_requirements_pex_filename = 'pytest-with-requirements.pex'
-  requirements_pex_argv = [
-    python_binary,
-    './{}'.format(pex_snapshot.files[0]),
-    '-e', 'pytest:main',
-    '-o', output_pytest_requirements_pex_filename,
-  ] + interpreter_constraint_args + [
-    # TODO(#7061): This text_type() wrapping can be removed after we drop py2!
-    text_type(req) for req in all_requirements
-  ]
-  pex_resolve_env = {'PATH': interpreter_search_paths}
-  # TODO(#6071): merge the two dicts via ** unpacking once we drop Py2.
-  pex_resolve_env.update(pex_build_environment.invocation_environment_dict)
-  requirements_pex_request = ExecuteProcessRequest(
-    argv=tuple(requirements_pex_argv),
-    env=pex_resolve_env,
-    input_files=pex_snapshot.directory_digest,
-    description='Resolve requirements: {}'.format(", ".join(all_requirements)),
-    output_files=(output_pytest_requirements_pex_filename,),
-  )
   requirements_pex_response = yield Get(
```

Comment:
Main blocking question for this PR. Is this even possible to `yield Get` a rule that has multiple parameters? It looks like this is all blocked by https://github.com/pantsbuild/pants/issues/7490.

Reply:
I would suggest having both input and output datatypes for this, which would resolve that issue, and be more typesafe than requesting `ExecuteProcessResult` directly.
```
yield Get(RequirementsPex, RequirementsPexRequest(all_requirements, output_pytest_requirements_pex_filename, "pytest:main", interpreter_constraints))
```